GVFS/GVFS.Common/PlaceholderListDatabase.cs

```diff
@@ -167,6 +167,37 @@ public void GetAllEntries(out List<PlaceholderData> filePlaceholders, out List<P
             }
         }
 
+        public void GetAllFileEntries(out Dictionary<string, PlaceholderListDatabase.PlaceholderData> filePlaceholdersByPath)
+        {
+            try
+            {
+                Dictionary<string, PlaceholderListDatabase.PlaceholderData> filePlaceholdersFromDiskByPath = 
+                    new Dictionary<string, PlaceholderListDatabase.PlaceholderData>(Math.Max(1, this.EstimatedCount), StringComparer.Ordinal);
+
+                string error;
+                if (!this.TryLoadFromDisk<string, string>(
+                    this.TryParseAddLine,
+                    this.TryParseRemoveLine,
+                    (key, value) =>
+                    {
+                        if (value != PartialFolderValue && value != ExpandedFolderValue)
+                        {
+                            filePlaceholdersFromDiskByPath[key] = new PlaceholderData(path: key, fileShaOrFolderValue: value);
+                        }

```

Comment:
@wilbaker to make sure my understanding is correct his returns all the files (instead of both files and directories).

Did we decide against just doing something like:
   var AllEntries = GetAllEntries();
   var results = AllEntries.where(x => x.value != PartialFolderValue && x.value != ExpandedFolderValue);
Due to performance reasons?

Reply:
@wilbaker do we do anything to enforce that second point? Otherwise having both a `GetAllEntries` and `GetAllFileEntries` is going to lead to a bug down the road when someone calls the wrong one.

But also, I don't quite understand the issue. What would happen if we call `GetAllEntries` without calling `WriteAllEntriesAndFlush`? I don't see `GetAllEntries` changing any state, so I would not have known not to call it in this situation.

Reply:
> I don't see GetAllEntries changing any state, so I would not have known not to call it in this situation.

The state change and enforcement in `GetAllEntries` is here:

```
() =>
{
    if (this.placeholderDataEntries != null)
    {
        throw new InvalidOperationException("PlaceholderListDatabase should always flush queue placeholders using WriteAllEntriesAndFlush before calling GetAllEntries again.");
    }

    this.placeholderDataEntries = new List<PlaceholderDataEntry>();
}))
```
> do we do anything to enforce that second point?

If you were to call `GetAllEntries` without calling `WriteAllEntriesAndFlush`, whenever the next call to `GetAllEntries` happens (e.g. during a projection change) it will throw.

> But also, I don't quite understand the issue.

At a high level, the issue is that the `GetAllEntries` API was specifically tailored for the "update projection" scenario.  In that scenario `GetAllEntries` and `WriteAllEntriesAndFlush` are always paired.

In this new scenario we just want to delete some entries from the placeholder list (and not write them all back) and so we can't use the `GetAllEntries` getter.

What would you think if for now I renamed `GetAllEntries` to something like `GetAllEntriesAndPrepForWriteAllEntries`?  It's a bit of a mouthful but makes it clear this is not a generic getter.

Reply:
Thanks, I see it now, though I still can't claim I understand what's happening :-). (Also I just found the comments on the `placeholderDataEntries` variable, but even that doesn't explain it - _why_ are these entries going to be lost?)

Given that the method throws if you call it incorrectly, that's at least some protection. I'm not sure `PrepForWriteAllEntries` explains what's happening, and I don't yet have a better suggestion. I'll keep trying to digest what's happening though.

Reply:
> but even that doesn't explain it - why are these entries going to be lost?

I had to refresh myself on all this code when I made this change, and I was able to track the "lost entry" scenario down:

1. VFS4G parses the updated index (as part of a projection change)
2. VFS4G starts the work to update placeholders
3. VFS4G calls `GetAllEntries`
4. VFS4G starts updating placeholders
5. Some application reads a pure-virtual file (creating a new placeholder) while VFS4G is updating existing placeholders. (**That new placeholder is added to `placeholderDataEntries`**)
6. VFS4G completes updating the placeholders and calls `WriteAllEntriesAndFlush`. (**Note:** this list does *not* include the placeholders created in step 5, as the were not included in `GetAllEntries`).
7. `WriteAllEntriesAndFlush` writes *both* the entires in `placeholderDataEntries` and those that were passed in as the parameter.

Without `placeholderDataEntries` VFS4G would lose track of the placeholder creations that crossed with updating placeholders.

Reply:
OK thanks, that makes sense. I would consider adding that explanation in the code, and possibly rename the variable to something like `placeholdersCreatedDuringProjectionUpdate`. As for the method, how about `GetAllEntriesForProjectionUpdate`?

I know this was a detour from the point of this PR, but it was useful for understanding what's happening. Thanks for digging up those details.

Reply:
Thanks for the long explanation @wilbaker!  I have a better understanding of what's going on now.

Reply:
@wilbaker would it be possible and safer to have a method that takes a delegate like
`Func<List<PlaceholderListDatabase.PlaceholderData>, IEnumerable<PlaceholderData> updatedPlaceholders>`
Then we could hide the fact that these need to be tied together by having the method be something like
```
void UpdatePlaceholders(Func<List<PlaceholderListDatabase.PlaceholderData>, IEnumerable<PlaceholderData> updatedPlaceholders> updateWork)
{
    List<PlaceholderListDatabase.PlaceholderData> placeholders = this.GetAllEntries();
    this.WriteAllEntriesAndFlush(updateWork(placeholders));
}
```
And make the 2 methods private.

Reply:
>would it be possible and safer to have a method that takes a delegate like
Func<List<PlaceholderListDatabase.PlaceholderData>, IEnumerable<PlaceholderData> updatedPlaceholders>
Then we could hide the fact that these need to be tied together by having the method be something like

@kewillford I think that would be possible and safer, but I think the changes in this PR have become complicated enough that I'd rather not also include refactoring of code (beyond renaming) unrelated to the changes required to fix the issue this PR is addressing.
 
I've filed #426 to make the cleanup you're suggesting.